util/memory/tracker.go

```diff
@@ -301,28 +306,47 @@ func (t *Tracker) AttachToGlobalTracker(globalTracker *Tracker) {
 	if !globalTracker.isGlobal {
 		panic("Attach to a non-GlobalTracker")
 	}
-	if t.parent != nil {
-		if t.parent.isGlobal {
-			t.parent.Consume(-t.BytesConsumed())
+	parent := t.getParent()
+	if parent != nil {
+		if parent.isGlobal {
+			parent.Consume(-t.BytesConsumed())
 		} else {
-			t.parent.remove(t)
+			parent.remove(t)
 		}
 	}
-	t.parent = globalTracker
-	t.parent.Consume(t.BytesConsumed())
+	t.setParent(globalTracker)
+	globalTracker.Consume(t.BytesConsumed())
 }
 
 // DetachFromGlobalTracker detach itself from its parent
 // Note that only the parent of this tracker is Global Tracker could call this function
 // Otherwise it should use Detach
 func (t *Tracker) DetachFromGlobalTracker() {
-	if t.parent == nil {
+	parent := t.getParent()
+	if parent == nil {
 		return
 	}
-	if !t.parent.isGlobal {
+	if !parent.isGlobal {
 		panic("Detach from a non-GlobalTracker")
 	}
-	parent := t.parent
 	parent.Consume(-t.BytesConsumed())
-	t.parent = nil
+	t.setParent(nil)
+}
+
+// ReplaceBytesUsed replace bytesConsume for the tracker
+func (t *Tracker) ReplaceBytesUsed(bytes int64) {
+	t.Consume(-t.BytesConsumed())
+	t.Consume(bytes)
+}
+
+func (t *Tracker) getParent() *Tracker {
+	t.parMu.Lock()
+	defer t.parMu.Unlock()
```

Comment:
Why don't we lock the parent in functions like `Consume`, instead, to lock it here?

Reply:
Lock `parent` is my first trial when I found the data race error, maybe add lock in the `Consume` function is also enough. But the tpcc test result seems ok if we lock `parent` here.